### DISCUSSION TITLE
Send additional parameters on SignIn() with GoogleProvider

### DISCUSSION THREAD

**Discussion by zanzlender** [nextauthjs/next-auth#4094] [Help] [answered]
I have a special use case for my login.

Quick run down:
- I have 2 login forms for Users/Admins which are on the same page
- The auth logic for these 2 roles is different 
- I want credentials and Google login

Because of the fact that i have 2 forms and 2 roles, I want to send the user's role (depending on which form the user sent) together with the SignIn() request. 
With credentials provider I can do something like this:
```
signIn("credentials", {
      redirect: false,
      password: "password",
      email: "email@gmail.com",
      userType: "coach",
    });
```
and all these properties appear in the authorize callback inside CredentialsProvider. ✔

But now I want to do the same thing with GoogleProvider. In the docs I found only **signIn("google")**, but i tried it like credentials just to see. Which looks something like this: 
```
signIn("google", { 
      userType: "user" 
    });
```

Although it works like this, no matter what I send, somewhere along the way these additional properties (userType) get lost.... And I can't access them inside [..nextauth].ts. I also tried to override the **profile** but still can't access these additional parameters from request like with CredentialsProvider.

Anyone has some idea how I could do this? 🤯

**Comment by zanzlender** [comment, marked as answer]
**UPDATE**

I found the answer, in case anyone encounters this as well.


As per the docs [Next-Auth Additional Parameters](https://next-auth.js.org/getting-started/client#additional-parameters) I added my parameter inside Google's authorization parameter, like so: 

```
GoogleProvider({
        clientId: process.env.GOOGLE_CLIENT_ID,
        clientSecret: process.env.GOOGLE_CLIENT_SECRET,
        authorization: {
          params: {
            prompt: "consent",
            access_type: "offline",
            response_type: "code",
            userType: "user" || "admin",  <-- THIS ONE
          },
        },
      }),
```

The problem was that the additional parameters don't arrive in the signIn callback, where I run my logic, and I didn't feel like creating my own file for Google's callback. For a reminder I send them like so: 

```
signIn("google", { userType: "user" })
```

Because of that I tried using the OG request before it enters the Google OAuth cycle, and save that parameter into a global variable, like so:  
```
let userType = "";

export default async function auth(req: NextApiRequest, res: NextApiResponse) {
      // Do whatever you want here, before the request is passed down to `NextAuth`
      if (req.body.userType !== undefined) {
             userType = req.body.userType;
      };
  }
```

And it works! 🤯 However since this is triggered on each request sent between Next-auth and Google, we need to check if the userType parameter is not undefined, so we don't overwrite it. After that I can use it in any callback like this:

```
callbacks: {
      async signIn({ user, account, profile, email, credentials }) {
          if (account.provider === "google") {
                if (profile.email_verified && profile.email.endsWith("@gmail.com")) {
                       if (userType === "user") { }   <-- THIS
                }
         }
     },
}

```

**Reply by VictorKolb** [reply]
But this is wrong solution. In case if started two or more authentication you can't indicate user, because you save last `userType` in global scope. In few concurrent requests it will give last value whatever user is `admin` or `user`. It make sense if you have very small  users amount,. In this case the best solution is hardcoded admin emails.

**Reply by jaekunchoi** [reply]
how can we pass any additional params for google auth and retrieve it in callback?

**Reply by VictorKolb** [reply]
> how can we pass any additional params for google auth and retrieve it in callback?

The only solution I found is place the additional params into cookie and after successful authentication grab them from cookie :(

**Reply by derins** [reply]
Hey Victor, do you have any code examples for this?

**Reply by jaekunchoi** [reply]
@VictorKolb how do you get cookie?

**Reply by derins** [reply]
I was looking to use the parameters as callback, figured out the issue by ```signIn('discord', { callbackUrl: `/anything` })```
. Will take {anything} as a query parameter

**Reply by VictorKolb** [reply]
> Hey Victor, do you have any code examples for this?
> @VictorKolb how do you get cookie?

@derins @zanzlender 
hey guys! 
You can get cookie from `req`

`src/pages/api/auth/[...nextauth].ts`
```js
function auth(
  req: NextApiRequest,
  res: NextApiResponse,
): ReturnType<typeof NextAuth> {
  return NextAuth(req, res, nextAuthOptions(req, res));
}


function nextAuthOptions(
  req: NextApiRequest,
  res: NextApiResponse,
): NextAuthOptions {
  return {
    providers: [
      GoogleProvider({
        clientId: process.env.GOOGLE_ID || '',
        clientSecret: process.env.GOOGLE_SECRET || '',
      }),
      FacebookProvider({
        clientId: process.env.FACEBOOK_CLIENT_ID || '',
        clientSecret: process.env.FACEBOOK_CLIENT_SECRET || '',
      }),
      AppleProvider({
        clientId: process.env.APPLE_ID || '',
        clientSecret: process.env.APPLE_SECRET || '',
      }),
    ],
    callbacks: {
      async signIn({ account }): Promise<string> {
        try {
           //  your cookies and custom logic
           console.log(req.cookies)
   
          return '/';
        } catch (authError) {
          return Promise.reject(authError);
        }
      },
    },
    secret: 'secret',
    pages: {
      signIn: '/auth/signin/',
    },
  };
}

export default withSentry(auth);

```

**Reply by jaekunchoi** [reply]
great it works! How do you remove cookie then?

**Reply by VictorKolb** [reply]
> great it works! How do you remove cookie then?

@jaekunchoi 
something like this

```js
res.setHeader('Set-Cookie', [
  'COOKIENAME=;Path=/;EXPIRES=Thu, 01 Jan 1970 00:00:00 GMT',
  'COOKIENAME2=;Path=/;EXPIRES=Thu, 01 Jan 1970 00:00:00 GMT',
]);
```

**Reply by jaekunchoi** [reply]
nice one! thanks!

**Reply by nimeshvaghasiya** [reply]
> > how can we pass any additional params for google auth and retrieve it in callback?
> 
> The only solution I found is place the additional params into cookie and after successful authentication grab them from cookie :(


@VictorKolb How to set additional params into cookie, is it 'cookies-next' or nextjs provide any default syntax to set cookie?
I have set it inside getServerSideProps as below but while accessing(console.log(req.cookies)) it inside signIn() callback of next auth, it has nothing like 'register'.
context.res.setHeader('set-cookie', ['register=true']);

**Reply by VictorKolb** [reply]
@nimeshvaghasiya you need to double check getServerSideProps called before signIn() and cookie was set to browser. You can do it on Network tab in devtools. Select your request and check cookie tab.

**Reply by nimeshvaghasiya** [reply]
@VictorKolb I can confirm It hits getServerSideProps, and sets register = true in browser network tab. Then I have redirected to discord oauth provider on button click, then It is not accessible into signin() callback(...nextauth.ts) using req.cookies.

**Reply by nimeshvaghasiya** [reply]
If I set it up using [cookies-next](https://www.npmjs.com/package/cookies-next) > setCookie('register', 'true'); I get the register cookie in signin() callback. I'm unsure why it isn't available when I have set it using `context.res.setHeader('set-cookie', ['register=true']);`

**Reply by VictorKolb** [reply]
@nimeshvaghasiya Looks strange. Think you need to debugging this

**Reply by mkilincaslan** [reply]
thanks @zanzlender this is savior answer! 🙌

**Comment by arisAlexis** [comment]
setting a cookie to delete later in the process is really a weird way of doing this. Since next-auth provides a way to send these params to the api backend there should be a much simpler way to retrieve them.

**Reply by derins** [reply]
I think using a query parameter is a more convenient way

**Reply by VictorKolb** [reply]
Please let us know if you figure out how to do this.

**Reply by khuezy** [reply]
How would you get the parameter (from `signIn('google', {}, { parameter: 'hi' }`) to the `jwt` callback? The parameter gets sent to google oauth's endpoint but it's missing from the auth callback query.

**Reply by ofir-shapira-como** [reply]
@khuezy did you manage to get the parameter in the JWT callback?

**Reply by khuezy** [reply]
Nope, I ended up using cookies that expired in a minute to pass data through

**Reply by shamoons** [reply]
How did you set the cookie?

**Reply by khuezy** [reply]
@shamoons on the client side: `document.cookie = "thing=value; path=/; expires=<UTC date string>"`
in the `jwt callback`, you can use `cookies().set('thing', '', {expires: 0})` to immediately remove it.

**Comment by davidreis97** [comment]
There really should be a better way to do this than cookies. I use the EmailProvider so it's not uncommon for mobile users to start account creation on one browser and then click the login link on their email which opens an entirely different browser, which breaks the cookies hack suggested by @VictorKolb.

**Comment by djs42012** [comment]
Anyone have updates to share here? Found myself down this rabbit hole last night and concluded cookies are still the only solution. Is there a technical reason why the query parameters aren't available in the callback? It already seems weird enough that we can't pass anything to the `signIn()` call to post directly to a database via one of the adapters. I'm just here because I want to store more info upon account creation than nextauth provides functionality for. It all seems so strange that I feel like I'm missing something.

**Comment by huychau** [comment]
One more solution, just get the data from request by add the specific params to the `redirectUrl`. We can get the param values by accessing `req.cookies?.['next-auth.callback-url']`.

```
await signIn('google', {redirectUrl: '/?userType=user'})
```

In `[...nextauth].ts`
```
export default async function auth(req: NextApiRequest, res: NextApiResponse) {
  return await NextAuth(req, res, {
    providers: [
      GoogleProvider({
        clientId: process.env.GOOGLE_CLIENT_ID || '',
        clientSecret: process.env.GOOGLE_CLIENT_SECRET || '',
      }),
    ],
  
    callbacks: {
      async jwt(params) {
        const callbackUrl = req.cookies?.['next-auth.callback-url'] ||  req.cookies?.['__Secure-next-auth.callback-url']
        
        // Get the params from callbackUrl
      },
    }
  });
}
```

**Reply by khuezy** [reply]
But now the url has the param in it. Is there a way to remove it after grabbing the value from the cookie? This solution still uses cookies though.

**Comment by carrilhorafael57** [comment]
I had the same issue. I was actually retrieving the user from the db to pass some more information to the session and could not figure out a way to do that.

My solution was actually to make an api call inside the JWT callback. I am setting the token with the information that I need and passing it to the session:

`   async jwt({ token, account, user, trigger, session }) {

      // CHECK WHETHER THE USER LOGIN WITH PROVIDER OR CREDENTIALS TO SET THE TOKEN 
      if (account?.provider === "google" || account?.provider === "facebook") {
        const userData = await retrieveUser(account);

        token.user = userData;
        token.accessToken = account.access_token;
      } else if (account?.provider === "credentials") {
        token.user = user;
        token.accessToken = account.access_token;
      }

      if (trigger === "update") {
        token.user = session.user;
      }


      return token;
    },`

Hope it helps someone.

**Comment by drattansingh** [comment]
This isn't pretty but it seems to work. 

Client:
`signIn('google', { usertype: 'super administrator' })`

Server: [...nextauth].js

```
import fs from 'fs';
import path from 'path';
import {v4 as uuidv4} from 'uuid'

const tempfile=uuidv4()+'.txt' // generated unique name of the file
export default async function auth(request, response){
  // write the user type to a temporary file
  const filePath=path.join(process.cwd(), tempfile)
  if(request.body?.usertype!==undefined) fs.writeFileSync(filePath, request.body.usertype)
  
  return await NextAuth(request, response,{
	callbacks: {
		async jwt({ token, user, account }){
                        if(account?.provider==='google'){
                            // Read the client key from the tmp file and then delete the file
                            const filePath=path.join(process.cwd(), tempfile)
                            const usertype=fs.readFileSync(filePath, 'utf-8')
                            fs.unlinkSync(filePath)// delete the file
		            console.log(usertype) // --> this prints out : super administrator
                            .......
                        }
                return token;
            }
```

**Reply by khuezy** [reply]
Would not recommend since you're doing i/o

**Comment by dhruv9499** [comment]
still not proper solution guys ?

**Reply by davidreis97** [reply]
The solution:

https://auth0.com/
https://supabase.com/
https://clerk.com/

**Reply by drattansingh** [reply]
Using cookies seems to be the only viable way to do this

**Reply by xcaeser** [reply]
> The solution:
> 
> https://auth0.com/ https://supabase.com/ https://clerk.com/

lmao

**Comment by Tushant-Sharma** [comment]
This method is working fine for me . 

file : [...nextauth/route.tsx]
`let userType: string | null = "";

  async function auth(req: NextApiRequest, res: NextApiResponse) {

    const url = new URL(req.url || "");

    if (url.searchParams.get("usertype")) {
      userType = url.searchParams.get("usertype");
    }

    return await NextAuth(req, res, {
      ...NEXTAUTH_OPTIONS,
      callbacks: {
        async signIn({ user, account, profile, email, credentials }) {
          console.log("\n\n\n\n\n", userType, "\n\n\n\n\n");
          return "/dashboard";
        },
      },
    });
  `}``
export { auth as GET, auth as POST };

use the function to send the params : 

            `  await signIn('google', {
                callbackUrl: "/foo",
                redirect: true,
              }, {
              usertype: 'adminlikedin',
                scope: 'openid email profile',
              });`

**Reply by khuezy** [reply]
Don't do this, `userType` is at the top level scope and it will cause race conditions when you have multiple requests. This *WILL* cause issues in your application.

**Comment by pnutmath** [comment]
For Next AppRouter we can do something like this

````typescript
// app/api/auth/[...nextauth]/route.ts
import { getNextAuth } from "@/auth";
import { NextRequest, NextResponse } from "next/server";

async function customHandler(req: NextRequest, res: NextResponse) {
  if (!req.method) return;

  const { handlers }: { handlers: { [key: string]: any } } = getNextAuth(req);

  const response = await handlers[req.method](req, res);
  return response;
}

export { customHandler as GET, customHandler as POST };

````

and `getNextAuth` function as below

````typescript
// auth.ts
import NextAuth from "next-auth";
import { NextRequest } from "next/server";

export function getNextAuth(req: NextRequest): {
  handlers: { [key: string]: any };
} {
  return NextAuth({ ... })
}

````

**Reply by xcaeser** [reply]
> For Next AppRouter we can do something like this
> 
> 
> 
> ````typescript
> 
> // app/api/auth/[...nextauth]/route.ts
> 
> import { getNextAuth } from "@/auth";
> 
> import { NextRequest, NextResponse } from "next/server";
> 
> 
> 
> async function customHandler(req: NextRequest, res: NextResponse) {
> 
>   if (!req.method) return;
> 
> 
> 
>   const { handlers }: { handlers: { [key: string]: any } } = getNextAuth(req);
> 
> 
> 
>   const response = await handlers[req.method](req, res);
> 
>   return response;
> 
> }
> 
> 
> 
> export { customHandler as GET, customHandler as POST };
> 
> 
> 
> ````
> 
> 
> 
> and `getNextAuth` function as below
> 
> 
> 
> ````typescript
> 
> // auth.ts
> 
> import NextAuth from "next-auth";
> 
> import { NextRequest } from "next/server";
> 
> 
> 
> export function getNextAuth(req: NextRequest): {
> 
>   handlers: { [key: string]: any };
> 
> } {
> 
>   return NextAuth({ ... })
> 
> }
> 
> 
> 
> ````
> 
> 
> 
> 

never seen such a confusing reply. no explanation has been provided.
no example usage provided.
